File: gmail/gmail-auth.js
#!/usr/bin/env node
/**
 * Gmail OAuth Token Generator for OpenClaw
 * Generates and stores a token.json file for Gmail API access
 */

const fs = require('fs');
const path = require('path');
const { google } = require('googleapis');
const { authenticate } = require('@google-cloud/local-auth');

const SCOPES = ['https://www.googleapis.com/auth/gmail.readonly'];
const CREDENTIALS_FILE = path.join(__dirname, 'credentials.json');
const TOKEN_FILE = path.join(__dirname, 'token.json');

/**
 * Load or request or refresh saved credentials.
 */
async function loadSavedCredentialsIfExist() {
  try {
    const content = fs.readFileSync(TOKEN_FILE);
    const credentials = JSON.parse(content);
    return google.auth.fromJSON(credentials);
  } catch (err) {
    return null;
  }
}

/**
 * Serialize credentials to a file comptible with GoogleAUth.fromJSON.
 */
async function saveCredentials(client) {
  const content = fs.readFileSync(CREDENTIALS_FILE);
  const keys = JSON.parse(content);
  const key = keys.installed || keys.web;
  const payload = {
    type: 'authorized_user',
    client_id: key.client_id,
    client_secret: key.client_secret,
    refresh_token: client.credentials.refresh_token,
  };
  fs.writeFileSync(TOKEN_FILE, JSON.stringify(payload));
  console.log('Token saved to', TOKEN_FILE);
}

/**
 * Load or request or refresh saved credentials.
 */
async function authorize() {
  let client = await loadSavedCredentialsIfExist();
  if (client) {
    return client;
  }
  client = await authenticate({
    scopes: SCOPES,
    keyfilePath: CREDENTIALS_FILE,
  });
  if (client.credentials) {
    await saveCredentials(client);
  }
  return client;
}

async function main() {
  try {
    console.log('Starting Gmail OAuth authentication...');
    const auth = await authorize();
    console.log('✓ Authentication successful!');
    console.log('✓ Token saved to', TOKEN_FILE);
    console.log('\nYour Gmail API is now configured for OpenClaw.');
  } catch (err) {
    console.error('✗ Error:', err.message);
    process.exit(1);
  }
}

main();

File: gmail/gmail-auth.py
#!/usr/bin/env python3
"""
Gmail OAuth Token Generator for OpenClaw
Generates and stores a token.json file for Gmail API access
"""

import os
import pickle
import json
from google.auth.transport.requests import Request
from google.oauth2.credentials import Credentials
from google_auth_oauthlib.flow import InstalledAppFlow

SCOPES = ['https://www.googleapis.com/auth/gmail.readonly']
CREDENTIALS_FILE = 'credentials.json'
TOKEN_FILE = 'token.json'

def get_gmail_service():
    """Authenticate and return Gmail API service"""
    creds = None
    
    # Load existing token if it exists
    if os.path.exists(TOKEN_FILE):
        with open(TOKEN_FILE, 'rb') as token:
            creds = pickle.load(token)
    
    # If no valid credentials, run OAuth flow
    if not creds or not creds.valid:
        if creds and creds.expired and creds.refresh_token:
            creds.refresh(Request())
        else:
            flow = InstalledAppFlow.from_client_secrets_file(
                CREDENTIALS_FILE, SCOPES)
            creds = flow.run_local_server(port=0)
        
        # Save token for next time
        with open(TOKEN_FILE, 'wb') as token:
            pickle.dump(creds, token)
    
    return creds

if __name__ == '__main__':
    try:
        print("Starting Gmail OAuth authentication...")
        creds = get_gmail_service()
        print("✓ Authentication successful!")
        print(f"✓ Token saved to {TOKEN_FILE}")
        print("\nYour Gmail API is now configured for OpenClaw.")
    except Exception as e:
        print(f"✗ Error: {e}")
        exit(1)
